recursion/core-v2/src/lib.rs

```diff
@@ -1,167 +1,85 @@
 use std::iter::once;
 
-use p3_field::{Field, PrimeField32};
 use serde::{Deserialize, Serialize};
-use sp1_core::air::MachineProgram;
-use sp1_core::{air::PublicValues, stark::MachineRecord};
 use sp1_derive::AlignedBorrow;
-use sp1_recursion_core::{air::Block, runtime::D};
+use sp1_recursion_core::air::Block;
 
 pub mod alu_base;
 pub mod alu_ext;
 pub mod builder;
 pub mod machine;
 pub mod mem;
 pub mod program;
+pub mod runtime;
 
-// #[derive(Clone, Debug)]
-// pub struct Address;
-// pub type Address = u32;
+pub use runtime::*;
 
-// I don't think events should depend on the field being used,
-// but I don't want to implement encoding or memory yet
-#[derive(Clone, Debug)]
-pub struct BaseAluEvent<F> {
-    pub opcode: Opcode,
-    pub out: AddressValue<F, F>,
-    pub in1: AddressValue<F, F>,
-    pub in2: AddressValue<F, F>,
-    pub mult: F, // number of times we need this value in the future
-}
+#[derive(AlignedBorrow, Clone, Copy, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
+#[repr(C)]
+pub struct Address<F>(F);
 
-#[derive(Clone, Debug)]
-pub struct ExtAluEvent<F> {
-    pub opcode: Opcode,
-    pub out: AddressValue<F, Block<F>>,
-    pub in1: AddressValue<F, Block<F>>,
-    pub in2: AddressValue<F, Block<F>>,
-    pub mult: F, // number of times we need this value in the future
-}
+// -------------------------------------------------------------------------------------------------
 
-#[derive(Clone, Debug)]
-pub struct MemEvent<F> {
-    pub address_value: AddressValue<F, Block<F>>,
-    pub multiplicity: F,
-    pub kind: MemAccessKind,
+/// The inputs and outputs to an operation of the base field ALU.
+#[derive(AlignedBorrow, Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
+pub struct BaseAluIo<V> {
+    pub out: V,
+    pub in1: V,
+    pub in2: V,
 }
 
-#[derive(Clone, Debug)]
-pub enum MemAccessKind {
-    Read,
-    Write,
-}
-
-#[derive(AlignedBorrow, Clone, Copy, Debug, PartialEq, Eq)]
-#[repr(C)]
-/// A memory address along with the stored value.
-/// For alignment reasons, `val` is first --- in `AddressValue<F, Block<F>>`, `val` is well-aligned.
-pub struct AddressValue<A, V> {
-    val: V,
-    addr: A,
-}
+pub type BaseAluEvent<F> = BaseAluIo<F>;
 
-impl<A, V> AddressValue<A, V> {
-    pub fn new(addr: A, val: V) -> Self {
-        Self { addr, val }
-    }
+/// An instruction invoking the extension field ALU.
+#[derive(Clone, Debug, Serialize, Deserialize)]
```

Comment:
Something to think about, how to handle immediates? Will you just store to memory?